src/mondefense.cpp

```diff
@@ -1,38 +1,47 @@
+#include "mondefense.h"
 #include "monster.h"
+#include "creature.h"
+#include "damage.h"
 #include "game.h"
 #include "rng.h"
 #include "line.h"
 #include "bodypart.h"
-#include "speech.h"
 #include "messages.h"
-#include "mondefense.h"
-#include <algorithm>
 
-
-void mdefense::zapback(monster *m, Creature *source, const projectile *proj)
+void mdefense::zapback(monster *const m, Creature *const source, projectile const* const proj)
 {
-    player *foe = dynamic_cast< player* >( source );
-    monster *othermon = dynamic_cast< monster* >( source );
-    if( rl_dist( m->pos(), source->pos() ) > 1 ) {
-        return; // Out of range
+    if ((proj) ||                              // Not a melee attack
+        (rng(0, 100) > m->def_chance) ||       // Attacker lucked out
+        (rl_dist(m->pos(), source->pos()) > 1) // Out of range
+    ) {
+        return;
     }
-    if( proj != nullptr || rng(0, 100) > m->def_chance ) {
-        return; // Not a melee attack or attacker lucked out
+
+    auto const is_eligible_player = [](player const *const foe) {
```

Comment:
Scoping absolutely works for the problem at hand, bringing up shortcomings of it in other contexts is irrelevant.
There are lots of good places to use lambdas, the problem is you keep shoehorning them in everywhere else.
lambdas absolutely imply indirection, whether it can be optimized away is irrelevant, I don't care how fast it runs (this isn't remotely a hot path), what I care about is littering the code with lambdas for no good reason, or any other construct for that matter.